### PR TITLE
enable opt-in to redact Secret data from Object status

### PR DESCRIPTION
**Pull request by ravilr**
### Description of your changes

Fixes https://github.com/crossplane-contrib/provider-kubernetes/issues/178

- Adds a new flag `--sanitize-secrets`  when set in controllerConfig/deploymentRuntimeConfig of provider-kubernetes, redacts the Secret data from Object `status.atProvider.manifest`.

- the new redact behavior is off by default so as to not break any existing compositions relying on `status.atProvider.manifest.data` for patches.  Usages meant to shuttling observed secret data to/from remote/control-plane clusters can be updated to use  `spec.references[].patchesFrom` or `spec.connectionDetails` and avoid leaking secret data to Object MR status resources.

I have:

- [X] Read and followed Crossplane's [contribution process].
- [X] Run `make reviewable test` to ensure this PR is ready for review.

### How has this code been tested

- `make test` and `make e2e`

Updating the provider-kubernetes's deploymentRuntimeConfig with below additional arg:
```
         args:
           ......
+          - '--sanitize-secrets'
```

```
$ kubectl get provider.pkg provider-kubernetes
NAME                  INSTALLED   HEALTHY   PACKAGE                                           
provider-kubernetes   True        True      ravilr/provider-kubernetes:v0.11.6
```

will reconcile any existing and new Object resources managing Secrets to have their `status.atProvider.manifest.data` redacted:

```
apiVersion: kubernetes.crossplane.io/v1alpha2
kind: Object
metadata:
  annotations:
    crossplane.io/composition-resource-name: credential-secret
    crossplane.io/external-create-pending: "2024-02-07T07:29:20Z"
    crossplane.io/external-create-succeeded: "2024-02-07T07:29:20Z"
    crossplane.io/external-name: provider-token-nkm7s-9wlpl
  creationTimestamp: "2024-02-07T07:29:21Z"
  finalizers:
  - finalizer.managedresource.crossplane.io
  generateName: provider-token-nkm7s-
  generation: 2
  labels:
    crossplane.io/claim-name: provider-token
    crossplane.io/claim-namespace: awsuswest2-admin
    crossplane.io/composite: provider-token-nkm7s
  name: provider-token-nkm7s-9wlpl
  ownerReferences:
  - apiVersion: security.example.com/v1
    blockOwnerDeletion: true
    controller: true
    kind: XCredential
    name: provider-token-nkm7s
    uid: 1569a117-9227-4567-8db3-259e18171da5
  resourceVersion: "86940719"
  uid: 7ce9094d-f06a-4c9d-8e66-3f1a41544893
spec:
  deletionPolicy: Delete
  forProvider:
    manifest:
      apiVersion: v1
      kind: Secret
      metadata:
        name: provider-token
        namespace: awsuswest2-admin
      stringData:
        createdBy: provider-token-nkm7s
      type: Opaque
  managementPolicies:
  - '*'
  providerConfigRef:
    name: example
  readiness:
    policy: SuccessfulCreate
status:
  atProvider:
    manifest:
      apiVersion: v1
      data:
        redacted: null      #<<<<<<<<<======== this is the impact of this change.
      kind: Secret
      metadata:
        ......
      type: Opaque
  conditions:
  - lastTransitionTime: "2024-02-07T07:29:21Z"
    reason: Available
    status: "True"
    type: Ready
  - lastTransitionTime: "2024-02-07T07:29:21Z"
    reason: ReconcileSuccess
    status: "True"
    type: Synced
```

[contribution process]: https://git.io/fj2m9

### CODE REVIEW

**Comment by ravilr** [comment]
@lsviben @bobh66  PTAL.

File: internal/controller/object/object.go [hunk 1]
```diff
@@ -87,17 +88,19 @@ const (
 	errGetConnectionDetails = "cannot get connection details"
 	errGetValueAtFieldPath  = "cannot get value at fieldPath"
 	errDecodeSecretData     = "cannot decode secret data"
+	errSanitizeSecretData   = "failed sanitizing secret data"
```
Comment (turkenh) [inline internal/controller/object/object.go:91]:
```suggestion
	errSanitizeSecretData   = "cannot sanitizing secret data"
```

nit: for consistency with other errors.

File: internal/controller/object/object.go [hunk 2]
```diff
@@ -382,6 +388,16 @@ func getLastApplied(obj *v1alpha2.Object, observed *unstructured.Unstructured) (
 
 func (c *external) setObserved(obj *v1alpha2.Object, observed *unstructured.Unstructured) error {
 	var err error
+
+	if c.sanitizeSecrets {
+		if observed.GetKind() == "Secret" && observed.GetAPIVersion() == "v1" {
+			observed, err = sanitizeUnstructuredSecret(observed)
```
Comment (turkenh) [inline internal/controller/object/object.go:394]: We can leverage the [fieldpath](https://github.com/crossplane/crossplane-runtime/tree/master/pkg/fieldpath) library of crossplane-runtime instead, example: https://github.com/crossplane/function-sdk-go/blob/f466e52426a9316b99677a4d6409995e985b2c37/resource/composite/composite.go#L336

File: cmd/provider/main.go [hunk 3]
```diff
@@ -57,6 +57,7 @@ func main() {
 		leaderElection           = app.Flag("leader-election", "Use leader election for the controller manager.").Short('l').Default("false").Envar("LEADER_ELECTION").Bool()
 		maxReconcileRate         = app.Flag("max-reconcile-rate", "The number of concurrent reconciliations that may be running at one time.").Default("10").Int()
 		enableManagementPolicies = app.Flag("enable-management-policies", "Enable support for Management Policies.").Default("true").Envar("ENABLE_MANAGEMENT_POLICIES").Bool()
+		enableSanitizeSecrets    = app.Flag("enable-sanitize-secrets", "when enabled, redacts Secret data from Object status").Default("false").Envar("ENABLE_SANITIZE_SECRETS").Bool()
```
[the remark this answers is not in the document]
Comment (ravilr) [reply cmd/provider/main.go]: sure, updated to `--sanitize-secrets`

File: internal/controller/object/object.go (hunk 2 above)
Comment (ravilr) [reply internal/controller/object/object.go]: updated to use fieldpath library.

**Review by turkenh — APPROVED** [review]
Looking great, thanks @ravilr 🙌

**Comment by turkenh** [comment]
Yes, I think I forgot it in my [previous PR](https://github.com/crossplane-contrib/provider-kubernetes/pull/194). It would be great if you could do it here 🙏

**Comment by ravilr** [comment]
> Yes, I think I forgot it in my [previous PR](https://github.com/crossplane-contrib/provider-kubernetes/pull/194). It would be great if you could do it here 🙏

Done.

**Pull request merged**